### PR TITLE
Adding generic handler/formatter support to logging

### PR DESCRIPTION
**Pull request by oliverimc**
## Before this PR
If you want to do custom logging there is no way to integrate it first class with the compute module logging library, such that you get access to PID/Job-ID context variables. 

## After this PR
Adds generic support for handlers /formatters so you can integrate whatever logging behaviour like and SLS with a few imports of the library

## Are Docs needed?
Yes will need an example for SLS given that will be the major usage pattern

### CODE REVIEW

File: compute_modules/logging/common.py [hunk 1]
```diff
@@ -34,17 +34,24 @@
 
 
 # TODO: support for log file output (need access to selected log output location)
-def _create_logger(name: str) -> logging.Logger:
+def _create_logger(
+    name: str,
+    handlers: Optional[Union[logging.Handler, Tuple[logging.Handler, ...]]],
```
Comment (oliverimc) [inline compute_modules/logging/common.py]: I have added support for custom handlers, don't know if this is overkill but wanted to make it general (aligned with how sls does it) can scope it to only formatters if needed

File: compute_modules/logging/common.py [hunk 2]
```diff
@@ -66,10 +73,12 @@ class ComputeModulesLoggerAdapter(_LoggerAdapter):
     def __init__(
         self,
         logger_name: str,
+        handlers: Optional[Union[logging.Handler, Tuple[logging.Handler, ...]]],
+        formatter: Optional[logging.Formatter],
         process_id: int = -1,
         job_id: str = "",
     ) -> None:
-        self._p_logger = _create_logger(logger_name)
+        self._p_logger = _create_logger(logger_name, handlers, formatter)
```
Comment (oliverimc) [inline compute_modules/logging/common.py]: I "think" this should work with the way you have set up your adapter but will run some tests to check. Interested if you see any major immediate issues

**Review by oliverimc — COMMENTED** [review]
Added some initial comments, keen for thoughts before I start diving in with tests etc

File: compute_modules/logging/common.py [hunk 3]
```diff
@@ -27,24 +27,42 @@
...
+
+
 # TODO: add replica ID to default log format
 DEFAULT_LOG_FORMAT = (
     "%(levelname)-8s PID: %(process_id)-6s JOB: %(job_id)-36s LOC: %(filename)s:%(lineno)d - %(message)s"
 )
 
 
+def _setup_logger(
+    handlers: Optional[Union[logging.Handler, Tuple[logging.Handler, ...]]], formatter: Optional[logging.Formatter]
```
Comment (scottkruyswyk) [inline compute_modules/logging/common.py]: Might want to set defaults of `None` here if a user only wants to set a `formatter` but not `handlers` and vice-versa

```suggestion
    handlers: Optional[Union[logging.Handler, Tuple[logging.Handler, ...]]] = None, 
    formatter: Optional[logging.Formatter] = None,
```

File: compute_modules/logging/internal.py [hunk 4]
```diff
@@ -28,15 +28,22 @@ def set_internal_log_level(level: Union[str, int]) -> None:
...
-        )
+        refresh_internal_logger()
+
     return INTERNAL_LOGGER_ADAPTER
 
 
+def refresh_internal_logger():
+    global INTERNAL_LOGGER_ADAPTER
+    # Refresh the internal logger to match the format/handlers currently globally specified
+    COMPUTE_MODULES_ADAPTER_MANAGER.adapters.pop("compute_modules_internal", None)
```
Comment (scottkruyswyk) [inline compute_modules/logging/internal.py]: Clever 👏

**Review by scottkruyswyk — COMMENTED** [review]
One small nit, otherwise looks good! Thanks for taking care of this

**Pull request closed without merging**